src/main/java/org/wildfly/security/auth/client/SetCredentialStoreReferenceAuthenticationConfiguration.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2016 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.auth.client;
+
+import java.io.IOException;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+
+import org.wildfly.common.Assert;
+import org.wildfly.security.auth.callback.CredentialCallback;
+
+/**
+ * Subset of {@link AuthenticationConfiguration} handling {@link CredentialStoreReference} configuration.
+ *
+ * @author <a href="mailto:pskopek@redhat.com">Peter Skopek</a>
+ */
+
+class SetCredentialStoreReferenceAuthenticationConfiguration extends AuthenticationConfiguration {
+
+    private final CredentialStoreReference credentialStoreReference;
+
+    SetCredentialStoreReferenceAuthenticationConfiguration(AuthenticationConfiguration parent, CredentialStoreReference credentialStoreReference) {
+        super(parent);
+        Assert.checkNotNullParam("credentialStoreReference", credentialStoreReference);
+        this.credentialStoreReference = credentialStoreReference;
+    }
+
+    @Override
+    AuthenticationConfiguration reparent(AuthenticationConfiguration newParent) {
+        return new SetCredentialStoreReferenceAuthenticationConfiguration(newParent, credentialStoreReference);
+    }
+
+    @Override
+    void handleCallback(final Callback[] callbacks, final int index) throws IOException, UnsupportedCallbackException {
+        final Callback callback = callbacks[index];
+        if (callback instanceof CredentialCallback) {
+            return;
+        }
+        super.handleCallback(callbacks, index);
+    }
+
+    @Override
+    void handleCallbacks(final AuthenticationConfiguration config, final Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+        for (Callback callback : callbacks) {
+            if (callback instanceof CredentialCallback) {
+                CredentialCallback credentialCallback = (CredentialCallback) callback;
+                if (credentialCallback.getCredentialStoreReference() == null) {
+                    credentialCallback.setCredentialStoreReference(credentialStoreReference);
```

Comment:
I see, you are using the Callback to transfer the CredentialStoreReference from one AC to the other.

A problem here is you will be very dependent on the ordering the configuration is called to ensure the handleCallbacks method are called in the expected order.  As the handleCallbacks method receives the AuthenticationConfiguraton instance it may be better to just add a getCredentialStoreReference method to the API then your other implementation can make use of that.

On a related point maybe we should adjust the handleCallback method to also take the AuthenticationConfiguration as an argument, that way you can handle the CredentialCallback in that method without having to search the array.